waspc/src/Wasp/Generator/DbGenerator/Jobs.hs

```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
-  let npxPrismaCmd = ["npx", "prisma", "migrate", "dev", "--schema", SP.toFilePath schemaFile]
+  let npxPrismaMigrateCmd = npxPrismaCmd ++ ["migrate", "dev", "--schema", SP.toFilePath schemaFile]
   let scriptArgs =
         if System.Info.os == "darwin"
           then -- NOTE(martin): On MacOS, command that `script` should execute is treated as multiple arguments.
-            ["-Fq", "/dev/null"] ++ npxPrismaCmd
+            ["-Fq", "/dev/null"] ++ npxPrismaMigrateCmd
           else -- NOTE(martin): On Linux, command that `script` should execute is treated as one argument.
-            ["-feqc", unwords npxPrismaCmd, "/dev/null"]
+            ["-feqc", unwords npxPrismaMigrateCmd, "/dev/null"]
+
+  let job = runNodeCommandAsJob serverDir "script" scriptArgs J.Db
 
-  runNodeCommandAsJob serverDir "script" scriptArgs J.Db
+  retryJobOnErrorWith job (npmInstall projectDir) ForwardEverything
 
 -- | Runs `prisma studio` - Prisma's db inspector.
 runStudio :: Path' Abs (Dir ProjectRootDir) -> J.Job
 runStudio projectDir = do
   let serverDir = projectDir </> serverRootDirInProjectRootDir
   let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
 
-  runNodeCommandAsJob
-    serverDir
-    "npx"
-    [ "prisma",
-      "studio",
-      "--schema",
-      SP.toFilePath schemaFile
-    ]
-    J.Db
+  let npxPrismaStudioCmd = npxPrismaCmd ++ ["studio", "--schema", SP.toFilePath schemaFile]
+  let job = runNodeCommandAsJob serverDir (head npxPrismaStudioCmd) (tail npxPrismaStudioCmd) J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) ForwardEverything
+
+generatePrismaClient :: Path' Abs (Dir ProjectRootDir) -> J.Job
+generatePrismaClient projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+  let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
+
+  let npxPrismaGenerateCmd = npxPrismaCmd ++ ["generate", "--schema", SP.toFilePath schemaFile]
+  let job = runNodeCommandAsJob serverDir (head npxPrismaGenerateCmd) (tail npxPrismaGenerateCmd) J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) ForwardOnlyRetryErrors
+
+-- | Runs `npm install` to install dependencies.
+-- May be needed before npx commands if dependencies (e.g. Prisma) are not installed.
+npmInstall :: Path' Abs (Dir ProjectRootDir) -> J.Job
+npmInstall projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+  runNodeCommandAsJob serverDir "npm" ["install"] J.Db
+
+data JobMessageForwardingStrategy = ForwardEverything | ForwardOnlyRetryErrors
+
+data Attempt = InitialAttempt | Retry
+
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one last time.
+-- NOTE(shayne): We only want to forward one JobExit message to the external channel from this function, as callers will stop
+-- listening once they get it. This is why our job message forwarders do not send them and we do ourselves at the end.
```

Comment:
This is implementation details, so I would move this comment inside the function body, at the appropriate location. Caller of the function doesn't care about this, but if you are looking at the implementation, of it, then you do.

Reply:
This part with JobExit is particularly tricky but also important to understand because it is an assumption that is not easy to be aware of.

Therefore I think it makes sense to be very verbose here about it, even if too much, to ensure reader understands the reasoning. You could explain that what we are doing here is really creating a job that might run multiple sub-jobs as part of it. It might run just one sub-job, or it might run 3 sub-jobs (first attempt, recovery, second attempt). Since job signals its end by sending JobExit message to its channel, that means we must make sure to not send JobExit if only one of the multiple sub-jobs we are running here is done, therefore we can't just forward all of sub-job messages of we would also forward JobExit too early. Therefore we suppress their JobExit messages and make sure to send a JobExit only when we are done with all the sub-jobs.
You could explain this in even more details, focusing particularly on situation with 3 sub-jobs, and explaining how we don't want the first 2 of them to return their job exit.